### DISCUSSION TITLE
Proposal to make project directory the default place for installing ansible-lint requirements

### DISCUSSION THREAD

**Discussion by ssbarnea** [ansible/ansible-lint#2802] [Polls]
When running ansible lint in **online** mode and with the default `isolation: true`, where do you think it should install the requirements mentioned in requirements.yml file?

For those that do not know, the implicit isolation mode protects the user from surprises like having  his local list of collections or roles from `~/.ansible` accidentally polluted.

Also it ensures that working on one project does not affect of working on another project, aka cross-contamination. You can think about isolation mode as "containment", and it is implicit. Users can still disable it if they want.

The big question is what should be the ideal location to pick?

Please note that by `{project_dir}` we mean the folder that contains the `ansible.cfg` file, and also the CWD from which the linter runs by default. This option has the big advantage that it matches the behavior of ansible-playbook, which is automatically loading roles and collections adjacent to current folder, even if this folder is not mentioned in config file.

**Comment by briantist** [comment]
I could see utility in both the current and proposed locations. Would it be possible to choose on a case-by-case basis? Perhaps `isolation` changes to be multi-valued (`cache`, `project`, `config`, `none`), or a new option like `isolation_mode` that controls which method to use?

**Comment by felixfontein** [comment]
I don't think automatically installing collections into `{project_dir}/collections/` is a good idea. In some of my projects, I do have some local collections in there which are part of source control, and it would be really strange to have random other collections installed in that place just because I happened to run `ansible-lint`.

**Comment by drybjed** [comment]
I would like the option to use the first directory set in `ansible.cfg`. This allows for more complex repository layouts, where collections are not directly in the `collections/` subdirectory, but for example `ansible/collections/`. We could also be able to isolate the external dependencies to a separate ignored directory, for example `.cache/ansible/collections/` and deal with it that way.

**Comment by davedittrich** [comment]
Before `ansible-compat` was introduced, I used a very simple Linux symbolic link + environment variable method to handle this. It relied on existing (at the time) behavior of `molecule` and handled the situation describe above pretty well.  I [documented how to use it](https://github.com/davedittrich/utils/tree/v0.6.0#testing) and had `Makefile` rules to assist in setting things up.  Placing the secondary directory into a hidden `.cache` subdirectory is a little more opaque, but as long as the user is involved in the process they are more aware of ramifications.

When I came back to my code after a year or so, it completely broke as `ansible-compat` was doing all kinds of directory creation and deletion in an opaque way, in the `~/.cache` directory rather than the `~/.ansible` directory, and silently in the background, just to get practically the same results. It took me a while to figure out what was even going on.

I would strongly urge you *not* put extra things into the project repo directory, as that has implications for accidentally committing files to source repos, polluting backups with unwanted files, etc. Doing it in a non-transparent way to the user makes the problem even worse.

I'm sure something like the symbolic link method could be made more robust and implemented in other way besides a `Makefile`, but I agree with [felixfontein](https://github.com/felixfontein) that hidden side-effects can create a bad user experience.

**Comment by frittentheke** [comment]
I myself like the isolated path a little better, as it does not interfere with the other files.
But looking at this poll and all of the use-cases, it seems that there should simply be a switch / config option to "place" the collections downloaded by ansible-lint.

There are also "technical" reasons, e.g. when looking at Gitlab CI. They only allow add directories under the `$CI_PROJECT_DIR`, see the ancient issue https://gitlab.com/gitlab-org/gitlab/-/issues/14151.

Enabling Gitlab CI runners cache the requirements between pipeline runs will greatly speed those up. Currently one needs to add copy / move commands at the beginning and end of a job ... providing an options to simply have ansible-lint use a folder below the Ansible home would be awesome.

**Reply by ssbarnea** [reply]
That is indeed a very good argument towards moving the cache location inside the repository.

I am not in a hurry to implement these changes as I want to do more testing and see if we can do something that will allow ansible to find them without any extra configuration. The less dark-magic we use, the less surprises we will see from users of the linter.

**Reply by frittentheke** [reply]
Thanks for looking into this!
